libs/wingsdk/src/shared/sandbox.ts

```diff
@@ -109,55 +45,121 @@ export class Sandbox {
     }
   }
 
-  private loadBundleOnce() {
-    if (this.loaded) {
-      return;
-    }
-
-    if (!this.code) {
-      throw new Error("Bundle not created yet - please report this as a bug");
-    }
-
-    // this will add stuff to the "exports" object within our context
-    vm.runInContext(this.code!, this.context, {
-      filename: this.entrypoint,
-    });
-
-    this.loaded = true;
-  }
-
   public async call(fn: string, ...args: any[]): Promise<any> {
     // wait for the bundle to finish creation
     await this.createBundlePromise;
 
-    // load the bundle into context on the first run
-    // we don't do this earlier because bundled code may have side effects
-    // and we want to simulate that a function is "cold" on the first run
-    this.loadBundleOnce();
-
-    return new Promise(($resolve, $reject) => {
-      const cleanup = () => {
-        delete this.context.$resolve;
-        delete this.context.$reject;
-      };
-
-      this.context.$resolve = (value: any) => {
-        cleanup();
-        $resolve(value);
-      };
+    // emit an explicit error when trying to access __dirname and __filename because we cannot
+    // resolve these when bundling (this is true both for simulator and the cloud since we are
+    // bundling there as well).
+    const shim = `
+const { parentPort: $parent } = require('worker_threads');
+const $resolve = (value) => $parent.postMessage({ type: 'resolve', value });
+const $reject = (reason) => $parent.postMessage({ type: 'reject', reason });
+console.log = (...args) => $parent.postMessage({ type: 'log', args });
+console.debug = (...args) => $parent.postMessage({ type: 'debug', args });
+console.info = (...args) => $parent.postMessage({ type: 'info', args });
+console.warn = (...args) => $parent.postMessage({ type: 'warn', args });
+console.error = (...args) => $parent.postMessage({ type: 'error', args });
+process.exit = (exitCode) => $parent.postMessage({ type: 'exit', exitCode });
+Object.defineProperty(globalThis, "__dirname", {
+  get: () => { throw new Error("__dirname cannot be used within bundled cloud functions"); },
+});
+Object.defineProperty(globalThis, "__filename", {
+  get: () => { throw new Error("__filename cannot be used within bundled cloud functions"); },
+});
```

Comment:
Doesn't esbuild already fail if we have any __dirname or __filename?

Reply:
This is actually a major limitation we have today. There's a good chance that existing user code will depend on these.

Any thoughts on how to address this?

Reply:
@MarkMcCulloh From what I can tell esbuild doesn't reject code with __dirname, it just silently leaves it (leading to possibly wrong runtime behavior).

It looks like it could be possible to support it through an esbuild plugin? https://github.com/evanw/esbuild/issues/859